### ISSUE TITLE
Precision of results / using Kahan summation

### ISSUE THREAD

**Issue by msricher** [theochem/matrix-permanent#12]
The precision of the code isn't very high due to the number of operations done in a permanent evaluation.

We might be able to improve this by using [Kahan summation](https://en.wikipedia.org/wiki/Kahan_summation_algorithm) inside loops. It's worth trying.

**Comment by PaulWAyers** [comment]
Kahan summation also slows one down, so if the sum is large it will cost several times more CPU. [Pairwise summation](https://en.wikipedia.org/wiki/Pairwise_summation) may be a useful alternative. There is probably a C++ library out there for these.....

Admittedly, Kahan summation may be easier to implement; it adds some arithmetic operations but is easily implemented (and implemented in such a way that one can switch to it or not use it when needed) in the context of a normal `for` loop.

**Comment by msricher** [comment]
You're right, that should be good enough.  I don't know how much precision either will win, but it's worth trying sometime.

**Comment by PaulWAyers** [comment]
Kahan summation should, at least in its more refined versions, become independent of the number of samples. I'd guess that it would be roughly equal to the precision loss from summing the greatest and smallest numbers in the sum. The pairwise summation has an error that would be of the same order, but times $log_2(n)$ where $n$ is the number of terms summed; this would mean that the error would grow linearly with the size of the matrix instead of linearly with the number of terms in the sum, which I think is fine since it's intractable to take the permanent of large matrices anyway.
